In [1]:
#importer les packages
import numpy as np 
import pandas as pd
import matplotlib.pyplot as plt
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import LabelEncoder 
from sklearn.tree import DecisionTreeClassifier,plot_tree, export_text
from sklearn.metrics import accuracy_score
from sklearn import metrics
import seaborn as sns

# 1. Récupération des données 
L'objectif principal est de développer un modèle de prédiction capable 
d'Estimern l'approbation ou le rejet d'une demande de prêt hypothécairesion.

In [2]:
df=pd.read_csv('loan prediction.csv')

In [3]:
df.head()

Loan_ID Gender Married Dependents     Education Self_Employed  \
0  LP001002   Male      No          0      Graduate            No   
1  LP001003   Male     Yes          1      Graduate            No   
2  LP001005   Male     Yes          0      Graduate           Yes   
3  LP001006   Male     Yes          0  Not Graduate            No   
4  LP001008   Male      No          0      Graduate            No   

   ApplicantIncome  CoapplicantIncome  LoanAmount  Loan_Amount_Term  \
0             5849                0.0         NaN             360.0   
1             4583             1508.0       128.0             360.0   
2             3000                0.0        66.0             360.0   
3             2583             2358.0       120.0             360.0   
4             6000                0.0       141.0             360.0   

   Credit_History Property_Area Loan_Status  
0             1.0         Urban           Y  
1             1.0         Rural           N  
2             1.0         Urban           Y  
3             1.0         Urban           Y  
4             1.0         Urban           Y

In [4]:
df.shape

(614, 13)

In [5]:
 df.dtypes

Loan_ID               object
Gender                object
Married               object
Dependents            object
Education             object
Self_Employed         object
ApplicantIncome        int64
CoapplicantIncome    float64
LoanAmount           float64
Loan_Amount_Term     float64
Credit_History       float64
Property_Area         object
Loan_Status           object
dtype: object

In [6]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 614 entries, 0 to 613
Data columns (total 13 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   Loan_ID            614 non-null    object 
 1   Gender             601 non-null    object 
 2   Married            611 non-null    object 
 3   Dependents         599 non-null    object 
 4   Education          614 non-null    object 
 5   Self_Employed      582 non-null    object 
 6   ApplicantIncome    614 non-null    int64  
 7   CoapplicantIncome  614 non-null    float64
 8   LoanAmount         592 non-null    float64
 9   Loan_Amount_Term   600 non-null    float64
 10  Credit_History     564 non-null    float64
 11  Property_Area      614 non-null    object 
 12  Loan_Status        614 non-null    object 
dtypes: float64(4), int64(1), object(8)
memory usage: 62.5+ KB


In [7]:
df.isnull().sum()

Loan_ID               0
Gender               13
Married               3
Dependents           15
Education             0
Self_Employed        32
ApplicantIncome       0
CoapplicantIncome     0
LoanAmount           22
Loan_Amount_Term     14
Credit_History       50
Property_Area         0
Loan_Status           0
dtype: int64

Q1) à partir des codes  df.isnull().sum() & df.info() qu'est ce qu'on peut remarquer ?

Réponse :
D'après les codes df.isnull().sum() et df.info(), on peut remarquer les points suivants :

*** Le code df.isnull().sum() affiche le nombre de valeurs nulles dans chaque colonne du DataFrame. Dans ce cas, on peut voir par exemple que La colonne Loan_ID a 13 valeurs nulles, la colonne Gender a 3 valeurs nulles et la colonne Self_Employed a 32 valeurs nulles. Pour les autres colonnes n'ont aucune valeur nulle.

*** Le code df.info() affiche les informations sur le DataFrame, y compris les types de données de chaque colonne. Dans ce cas, on peut voir qu'il y a 8 colonnes d'objets, 1 colonne d'entiers et 4 colonnes de nombres flottants. De plus, le code df.info() affiche également la taille du DataFrame, qui est de 614 entrées. Cela signifie qu'il y a 614 observations dans le DataFrame.

La commande (df.describe()) nous permet d'avoir un résumé statistique sur nos données.

In [8]:
df.describe()

ApplicantIncome  CoapplicantIncome  LoanAmount  Loan_Amount_Term  \
count       614.000000         614.000000  592.000000         600.00000   
mean       5403.459283        1621.245798  146.412162         342.00000   
std        6109.041673        2926.248369   85.587325          65.12041   
min         150.000000           0.000000    9.000000          12.00000   
25%        2877.500000           0.000000  100.000000         360.00000   
50%        3812.500000        1188.500000  128.000000         360.00000   
75%        5795.000000        2297.250000  168.000000         360.00000   
max       81000.000000       41667.000000  700.000000         480.00000   

       Credit_History  
count      564.000000  
mean         0.842199  
std          0.364878  
min          0.000000  
25%          1.000000  
50%          1.000000  
75%          1.000000  
max          1.000000

In [9]:
 df.describe(include='O')

Loan_ID Gender Married Dependents Education Self_Employed  \
count        614    601     611        599       614           582   
unique       614      2       2          4         2             2   
top     LP001002   Male     Yes          0  Graduate            No   
freq           1    489     398        345       480           500   

       Property_Area Loan_Status  
count            614         614  
unique             3           2  
top        Semiurban           Y  
freq             233         422

Q2) Quelle est la différence entre les deux tableaux ?

Réponse :
On a def.describe() est utilisé pour les colonnes numériques, tandis que df.describe(include='O') est utilisé pour les colonnes de type objet. Ces deux méthodes permettent d'avoir avoir un résumé statistique sur nos données ous différentes perspectives.

In [10]:
 df.describe(include='all')

Loan_ID Gender Married Dependents Education Self_Employed  \
count        614    601     611        599       614           582   
unique       614      2       2          4         2             2   
top     LP001002   Male     Yes          0  Graduate            No   
freq           1    489     398        345       480           500   
mean         NaN    NaN     NaN        NaN       NaN           NaN   
std          NaN    NaN     NaN        NaN       NaN           NaN   
min          NaN    NaN     NaN        NaN       NaN           NaN   
25%          NaN    NaN     NaN        NaN       NaN           NaN   
50%          NaN    NaN     NaN        NaN       NaN           NaN   
75%          NaN    NaN     NaN        NaN       NaN           NaN   
max          NaN    NaN     NaN        NaN       NaN           NaN   

        ApplicantIncome  CoapplicantIncome  LoanAmount  Loan_Amount_Term  \
count        614.000000         614.000000  592.000000         600.00000   
unique              NaN                NaN         NaN               NaN   
top                 NaN                NaN         NaN               NaN   
freq                NaN                NaN         NaN               NaN   
mean        5403.459283        1621.245798  146.412162         342.00000   
std         6109.041673        2926.248369   85.587325          65.12041   
min          150.000000           0.000000    9.000000          12.00000   
25%         2877.500000           0.000000  100.000000         360.00000   
50%         3812.500000        1188.500000  128.000000         360.00000   
75%         5795.000000        2297.250000  168.000000         360.00000   
max        81000.000000       41667.000000  700.000000         480.00000   

        Credit_History Property_Area Loan_Status  
count       564.000000           614         614  
unique             NaN             3           2  
top                NaN     Semiurban           Y  
freq               NaN           233         422  
mean          0.842199           NaN         NaN  
std           0.364878           NaN         NaN  
min           0.000000           NaN         NaN  
25%           1.000000           NaN         NaN  
50%           1.000000           NaN         NaN  
75%           1.000000           NaN         NaN  
max           1.000000           NaN         NaN

Q3) Expliquer pour quoi on trouve des valeurs NAN dans ce résultats 

Réponse : 
Dans ces résultats, on trouve des valeurs NaN pour indiquer que certaines statistiques descriptives ne sont pas calculées ou ne sont pas pertinentes pour certaines colonnes en raison de leur type de données.

# 2. Nettoyage des données

In [11]:
#renseigner les valeurs manquantes
cat_data=[]
num_data=[]
for i,c in enumerate (df.dtypes):
     if c == 'object':
        cat_data.append(df.iloc[:,i])  # Ajoutez le nom de la colonne aux données catégorielles
     else:
        num_data.append(df.iloc[:,i])  # Ajoutez le nom de la colonne aux données numériques

In [12]:
cat_data=pd.DataFrame(cat_data).transpose()
cat_data

Loan_ID  Gender Married Dependents     Education Self_Employed  \
0    LP001002    Male      No          0      Graduate            No   
1    LP001003    Male     Yes          1      Graduate            No   
2    LP001005    Male     Yes          0      Graduate           Yes   
3    LP001006    Male     Yes          0  Not Graduate            No   
4    LP001008    Male      No          0      Graduate            No   
..        ...     ...     ...        ...           ...           ...   
609  LP002978  Female      No          0      Graduate            No   
610  LP002979    Male     Yes         3+      Graduate            No   
611  LP002983    Male     Yes          1      Graduate            No   
612  LP002984    Male     Yes          2      Graduate            No   
613  LP002990  Female      No          0      Graduate           Yes   

    Property_Area Loan_Status  
0           Urban           Y  
1           Rural           N  
2           Urban           Y  
3           Urban           Y  
4           Urban           Y  
..            ...         ...  
609         Rural           Y  
610         Rural           Y  
611         Urban           Y  
612         Urban           Y  
613     Semiurban           N  

[614 rows x 8 columns]

In [13]:
num_data=pd.DataFrame(num_data).transpose()
num_data

ApplicantIncome  CoapplicantIncome  LoanAmount  Loan_Amount_Term  \
0             5849.0                0.0         NaN             360.0   
1             4583.0             1508.0       128.0             360.0   
2             3000.0                0.0        66.0             360.0   
3             2583.0             2358.0       120.0             360.0   
4             6000.0                0.0       141.0             360.0   
..               ...                ...         ...               ...   
609           2900.0                0.0        71.0             360.0   
610           4106.0                0.0        40.0             180.0   
611           8072.0              240.0       253.0             360.0   
612           7583.0                0.0       187.0             360.0   
613           4583.0                0.0       133.0             360.0   

     Credit_History  
0               1.0  
1               1.0  
2               1.0  
3               1.0  
4               1.0  
..              ...  
609             1.0  
610             1.0  
611             1.0  
612             1.0  
613             0.0  

[614 rows x 5 columns]

Q4) Quel est l'objectif de ces trois codes? est-t-il necessaire de faire cette étape avant de gérer les données manquantes?

Réponce : 

L'objectif de ces trois codes :
** Pour gérer les valeurs manquantes dans les variables catégoriques
** Remplacer les valeurs manquantes par la valeur la plus fréquente (mode) est une approche raisonnable. 

Il necessaire de faire cette étape avant de gérer les données manquantes par ce que cela peut aider à préserver la distribution des données et à éviter d'introduire des biais.

Pour les variables categoriques manquantes on va  les remplacer par les valeurs qui se répetent le plus

Q5) est ce qu'il ya une autre alternative 

Réponce : 
Alors pour gérer les valeurs manquantes dans les variables catégoriques, il existe plusieurs alternatives : 
*** On peut remplacer les valeurs manquantes par la valeur la plus fréquente
*** Créer une nouvelle catégorie pour les valeurs manquantes, 

In [14]:
cat_data = cat_data.apply(lambda x: x.fillna(x.value_counts().index[0]))
cat_data.isnull().sum().any()

False

Vérification (on va prendre une seule variable catégorique pour tester le résultat avant et aprés la modifcation)

In [15]:
#avant
print (df['Gender'].value_counts())
print (df['Gender'].value_counts().sum())

Gender
Male      489
Female    112
Name: count, dtype: int64
601


In [16]:
#aprés
print (cat_data['Gender'].value_counts())
print (cat_data['Gender'].value_counts().sum())

Gender
Male      502
Female    112
Name: count, dtype: int64
614


Pour les variables numériques manquantes on va remplacer les variables manquantes par la valeur précedente de la meme colonne

In [17]:
num_data.fillna(method='bfill', inplace=True)
num_data.isnull().sum().any()

C:\Users\Ghofrane\AppData\Local\Temp\ipykernel_62964\1069030303.py:1: FutureWarning: DataFrame.fillna with 'method' is deprecated and will raise in a future version. Use obj.ffill() or obj.bfill() instead.
  num_data.fillna(method='bfill', inplace=True)


False

In [25]:
num_data.bfill(inplace=True)
num_data.isnull().sum().any()

False

Transformation de la colonne cible "Loan_Status" en une variable numérique et la nommer " target" puis supprimer la colone Loan_status
Pour supprimer une colonne, on utilise la commande "drop". Le paramètre axis=1 indique qu'on veut supprimer une colonne, et 'inplace=True' signifie que la modification doit être effectuée directement sur le DataFrame lui même.

In [18]:
target=cat_data['Loan_Status']
cat_data.drop('Loan_Status',axis=1,inplace=True)
target=target.map({'Y': 1, 'N': 0})
target

0      1
1      0
2      1
3      1
4      1
      ..
609    1
610    1
611    1
612    1
613    0
Name: Loan_Status, Length: 614, dtype: int64

Remplacer les valeurs categoriques par des valeurs numeriques

In [19]:
le=LabelEncoder()
for i in cat_data:
    cat_data[i]=le.fit_transform(cat_data[i])
cat_data

Loan_ID  Gender  Married  Dependents  Education  Self_Employed  \
0          0       1        0           0          0              0   
1          1       1        1           1          0              0   
2          2       1        1           0          0              1   
3          3       1        1           0          1              0   
4          4       1        0           0          0              0   
..       ...     ...      ...         ...        ...            ...   
609      609       0        0           0          0              0   
610      610       1        1           3          0              0   
611      611       1        1           1          0              0   
612      612       1        1           2          0              0   
613      613       0        0           0          0              1   

     Property_Area  
0                2  
1                0  
2                2  
3                2  
4                2  
..             ...  
609              0  
610              0  
611              2  
612              2  
613              1  

[614 rows x 7 columns]

Regrouper les données numérique et catégoriques

In [20]:
df=pd.concat([cat_data,num_data,target],axis=1)

In [21]:
df

Loan_ID  Gender  Married  Dependents  Education  Self_Employed  \
0          0       1        0           0          0              0   
1          1       1        1           1          0              0   
2          2       1        1           0          0              1   
3          3       1        1           0          1              0   
4          4       1        0           0          0              0   
..       ...     ...      ...         ...        ...            ...   
609      609       0        0           0          0              0   
610      610       1        1           3          0              0   
611      611       1        1           1          0              0   
612      612       1        1           2          0              0   
613      613       0        0           0          0              1   

     Property_Area  ApplicantIncome  CoapplicantIncome  LoanAmount  \
0                2           5849.0                0.0       128.0   
1                0           4583.0             1508.0       128.0   
2                2           3000.0                0.0        66.0   
3                2           2583.0             2358.0       120.0   
4                2           6000.0                0.0       141.0   
..             ...              ...                ...         ...   
609              0           2900.0                0.0        71.0   
610              0           4106.0                0.0        40.0   
611              2           8072.0              240.0       253.0   
612              2           7583.0                0.0       187.0   
613              1           4583.0                0.0       133.0   

     Loan_Amount_Term  Credit_History  Loan_Status  
0               360.0             1.0            1  
1               360.0             1.0            0  
2               360.0             1.0            1  
3               360.0             1.0            1  
4               360.0             1.0            1  
..                ...             ...          ...  
609             360.0             1.0            1  
610             180.0             1.0            1  
611             360.0             1.0            1  
612             360.0             1.0            1  
613             360.0             0.0            0  

[614 rows x 13 columns]

Vérifier la distribution des classes 

Q6) Qu'est ce qu'on peut remarquer? (Question relative à cette commande  : target.value_counts())

Réponce :
Après la transformation de la variable cible (Loan_Status) en une variable numérique nommée « target » et la suppression de cette colonne, on remarque une certaine disparité dans la répartition des classes au sein de la nouvelle variable « target ».
En effet, on observe 422 occurrences de la classe 1 (représentant les prêts approuvés) et 192 occurrences de la classe 0 (représentant les prêts rejetés). Cela signifie qu'il y a plus de prêts approuvés que de prêts rejetés dans l'ensemble de données.

In [22]:
target.value_counts()

Loan_Status
1    422
0    192
Name: count, dtype: int64

# Préparation des données pour l'apprentissage

Division des données en train et test 

random_state : une valeur entière qui détermine la reproductibilité des résultats. Fixer cette valeur garantit que chaque fois que vous exécutez le code, vous obtenez la même division des données.

stratify : spécifie la colonne à utiliser pour la stratification lors de la division des données. La stratification garantit que la répartition des classes dans l'ensemble d'entraînement et dans l'ensemble de test est similaire à celle du DataFrame d'origine, en particulier important pour les ensembles de données déséquilibrés.

In [23]:
dfTrain, dfTest = train_test_split(df, test_size=300, random_state=1, stratify=df.Loan_Status)
print(dfTrain.shape)
print(dfTest.shape)

(314, 13)
(300, 13)


In [24]:
print ("distribution relative des classes en apprentissage")
print (dfTrain.Loan_Status.value_counts(normalize=True))
print ("distribution relative des classes en test")
print (dfTest.Loan_Status.value_counts(normalize=True))

distribution relative des classes en apprentissage
Loan_Status
1    0.687898
0    0.312102
Name: proportion, dtype: float64
distribution relative des classes en test
Loan_Status
1    0.686667
0    0.313333
Name: proportion, dtype: float64


Q7) Qu'est ce qu'on peut remarquer